custom_components/tesla_custom/climate.py

```diff
@@ -52,125 +73,122 @@ def hvac_mode(self):
 
         Need to be one of HVAC_MODE_*.
         """
-        if self.tesla_device.is_hvac_enabled():
+        if self._car.is_climate_on:
             return HVAC_MODE_HEAT_COOL
+
         return HVAC_MODE_OFF
 
     @property
     def hvac_modes(self):
-        """Return the list of available hvac operation modes.
+        """Return list of available hvac operation modes.
 
         Need to be a subset of HVAC_MODES.
         """
         return SUPPORT_HVAC
 
     @property
     def temperature_unit(self):
-        """Return the unit of measurement."""
-        if self.tesla_device.measurement == "F":
-            return TEMP_FAHRENHEIT
+        """Return unit of measurement.
+
+        Tesla API always returns in Celsius.
+        """
         return TEMP_CELSIUS
 
     @property
     def current_temperature(self):
-        """Return the current temperature."""
-        return self.tesla_device.get_current_temp()
+        """Return current temperature."""
+        return self._car.inside_temp
+
+    @property
+    def max_temp(self):
+        """Return max temperature."""
+        if self._car.max_avail_temp:
+            return self._car.max_avail_temp
+
+        return DEFAULT_MAX_TEMP
+
+    @property
+    def min_temp(self):
+        """Return min temperature"""
+        if self._car.min_avail_temp:
+            return self._car.min_avail_temp
+
+        return DEFAULT_MIN_TEMP
 
     @property
     def target_temperature(self):
-        """Return the temperature we try to reach."""
-        return self.tesla_device.get_goal_temp()
+        """Return target temperature."""
+        return self._car.driver_temp_setting
 
-    @TeslaDevice.Decorators.check_for_reauth
     async def async_set_temperature(self, **kwargs):
-        """Set new target temperatures."""
+        """Set new target temperature."""
         temperature = kwargs.get(ATTR_TEMPERATURE)
         if temperature:
             _LOGGER.debug("%s: Setting temperature to %s", self.name, temperature)
-            await self.tesla_device.set_temperature(temperature)
-            self.async_write_ha_state()
 
-    @TeslaDevice.Decorators.check_for_reauth
+            temp = round(temperature, 1)
+
+            await self._car.set_temperature(temp)
+            # We'll create a non-blocking update call so we don't hold up
+            # the current call.
+            await self.update_controller(
```

Comment:
Calling `update_controller` isn't working as intended because Tesla's servers don't seem to be updating the state fast enough so we end up just pulling the old state.

Only thought is if we could add some sort of non-blocking delay to give Tesla's servers time to set the new state.

Reply:
The climate state appears to have this problem. See https://github.com/alandtse/tesla/issues/226

Reply:
So I think just awaiting a sleep would work, but we would just have to pick some value that gives enough time for Tesla's side to update the state before we force an update. I also don't know if updating all vehicle data is necessary if it's only one or two other states that change.

The other option is just to figure out what else is affected by this specific change and also "manually" update their states and just refresh all entities in HA. For example, I'm assuming we're forcing an update here because when you change the temperature, the climate system will turn on if it's off? If this is true, then on the teslajsonpy side I would also update the climate state in addition to the temperature state. As long as we know what else is changing, I think this might be a viable option and it saves us from having to poll the vehicle data endpoint every time.